pageserver/src/tenant/inmemory_layer.rs

```diff
@@ -356,17 +364,85 @@ impl InMemoryLayer {
 
         let mut keys: Vec<(&Key, &VecMap<Lsn, u64>)> = inner.index.iter().collect();
         keys.sort_by_key(|k| k.0);
-
+        let mut total_pages = 0usize;
+        let mut total_records = 0usize;
+        let mut collapsed_records = 0usize;
+        let mut generated_images = 0usize;
         for (key, vec_map) in keys.iter() {
             let key = **key;
             // Write all page versions
-            for (lsn, pos) in vec_map.as_slice() {
+            let page_updates = vec_map.as_slice();
+            let mut first_record = true;
+            let mut xid = 0u32;
+            let mut last_lsn: Lsn = Lsn::INVALID;
+            let mut records: Vec<(Lsn, ZenithWalRecord)> = Vec::new();
+            total_pages += 1;
+            total_records += page_updates.len();
+            for (lsn, pos) in page_updates {
                 cursor.read_blob_into_buf(*pos, &mut buf)?;
-                let will_init = Value::des(&buf)?.will_init();
-                delta_layer_writer.put_value_bytes(key, *lsn, &buf, will_init)?;
+                last_lsn = *lsn;
+                let value = Value::des(&buf)?;
+                let will_init = value.will_init();
+                // If sequence of WAL records for this page belongs to the same transaction, starts with initialization record
+                // and total size of records is larger than page size, then better materialize this page and store it as page image
```

Comment:
It's not that simple. There must also be no commit records in between, and you must coalesce *all* records between the collapsed range to a single LSN, not just the records on this page.

For example, consider this sequence of WAL records:

1. HEAP_INSERT row 'foo' (transaction A)
2. BTREE_INSERT ptr for row 'foo' (transaction A)
3. Commit transaction B
4. HEAP_INSERT row 'bar' (transaction A)
5. BTREE_INSERT ptr for row 'bar'

You cannot coalesce the HEAP_INSERT records from steps 1 and 4. If you did that, and you then did Point-in-Time query with the LSN from step 3, it would see the btree-insertion of row 'foo' but not the heap-insertion.

I think you need to do this earlier in the WAL ingestion pipeline, probably in walingest.rs, where you see all the incoming records.

Reply:
Sorry, do not completely understand.
Assume that we collapse two heap inserts of transaction A into one page image.
This page image is stored with LSN of second insert (but it is actually not important).
If you do PITR at point 3, then transaction A is not committed at this point and its results are not visible in any case. So we have page image containing records inserted by A, but them are not visible because transaction is not marked as committed in CLOG.

So my assumption is that it is always safe to collapse changes done within one transaction because Postgres ACID mechanism guarantee that either all this changes are visible either none of them are visible. You think that this assumption is wrong?

Reply:
You need to ensure the index is consistent with the heap. If the index tuple is present, the corresponding heap tuple must also be present.

Reply:
> You need to ensure the index is consistent with the heap. If the index tuple is present, the corresponding heap tuple must also be present.

 I see.
But does storing result page image at LSN of first collapsed WAL record solves this problem?
In this case referenced record will be always present in heap page.
Opposite case (when record is removed) is not possible in Postgres MVCC model.

Reply:
> > You need to ensure the index is consistent with the heap. If the index tuple is present, the corresponding heap tuple must also be present.
> 
> I see. But does storing result page image at LSN of first collapsed WAL record solves this problem? In this case referenced record will be always present in heap page. Opposite case (when record is removed) is not possible in Postgres MVCC model.

Then you would have the same problem if you coalesced a vacuum record. The vacuum record could remove a tuple from the heap page before the corresponding index tuple has been removed.

This is also totally unsafe with index pages, you mustn't see only one half of an page split, for example.

Reply:
> Then you would have the same problem if you coalesced a vacuum record. 

Then I can perform collapse only of heap INSERT and MULTIINSERT records.
It should solve the problem with vacuum and with index updates, shouldn't it?

Reply:
Actually I think that  writing frozen layer is the only place where this optimization can be performed.
Moving it to walingest caue the following problems:
1. We will have to perform walredo in walrecevier thread which can significantly slow down  it and trigger backpressure at compute node.
2. We will have to scatter data by transactions (as it is done for example in logical replication). And it will really "opens pandora's box", because maintaining of this decoding buffer cause a lot of problems which are not yet all fixed in logical replication.

Reply:
> Then I can perform collapse only of heap INSERT and MULTIINSERT records.
> It should solve the problem with vacuum and with index updates, shouldn't it?

I guess, but makes it pretty limited.

Hmm, I think it could cause confusion if have a read-only replica that fetches the page between those records. It would have one version of the page in cache with LSN X, and the primary would be have another version of the page with the same LSN and different contents.

You also mustn't coalesce records if a branch has been created between the first and last LSN. The page might've been updated on the branch, based on just the first insert on the page.